### PR TITLE
Temporarily fix for #2612

### PR DESCRIPTION
**Pull request by SamantazFox**
Don't rely on the auto compression/decompression provided by the crystal stdlib.

### CODE REVIEW

**Comment by unixfox** [comment]
Duplicating the comment from @rderensy here:

> Well, doesn't work now that https://github.com/iv-org/invidious/pull/2623 is installed, and refreshing doesn't solve it 
> 
> Title: `Unexpected char '' at line 1, column 1 (JSON::ParseException)`
> Date: `2021-11-16T18:48:38Z`
> Route: `/watch?v=dr1KJD_HAMs&quality=dash`
> Version: `2021.11.16-7129c3c6 @ master`
> <details>
> <summary>Backtrace</summary>
> <p>
>    
> ```
> Unexpected char '' at line 1, column 1 (JSON::ParseException)
>   from /usr/share/crystal/src/json/lexer.cr:331:7 in 'raise'
>   from /usr/share/crystal/src/json/lexer.cr:327:5 in 'unexpected_char'
>   from /usr/share/crystal/src/json/lexer.cr:326:11 in 'unexpected_char'
>   from /usr/share/crystal/src/slice.cr:213:5 in 'next_token'
>   from /usr/share/crystal/src/json/parser.cr:13:12 in '_post_json'
>   from src/invidious/yt_backend/youtube_api.cr:317:12 in 'extract_video_info:video_id'
>   from src/invidious/videos.cr:970:10 in 'get_video:region'
>   from src/invidious/routes/watch.cr:63:15 in 'handle'
>   from lib/kemal/src/kemal/route.cr:13:9 in '->'
>   from /usr/share/crystal/src/primitives.cr:266:3 in 'call'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
>   from lib/kemal/src/kemal/filter_handler.cr:21:7 in 'call'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
>   from src/invidious/helpers/static_file_handler.cr:167:15 in 'call'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call'
>   from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
>   from lib/kemal/src/kemal/init_handler.cr:12:7 in 'process'
>   from /usr/share/crystal/src/http/server.cr:500:5 in '->'
>   from /usr/share/crystal/src/primitives.cr:266:3 in 'run'
>   from ???
> ```
> </p>
> </details>

File: src/invidious/yt_backend/youtube_api.cr [hunk 1]
```diff
@@ -434,8 +425,23 @@ module YoutubeAPI
       )
     end
 
+    # Decompress the body ourselves, given that auto-decompress is
+    # broken in the Crystal stdlib.
+    # Read more:
+    #  - https://github.com/iv-org/invidious/issues/2612
+    #  - https://github.com/crystal-lang/crystal/issues/11354
+    #
+    case headers["Content-Encoding"]?
```
Comment (SamantazFox) [inline src/invidious/yt_backend/youtube_api.cr:434]:
```suggestion
    case response.headers["Content-Encoding"]?
```

**Comment by SamantazFox** [comment]
@rderensy Shoudl work better now xD

File: src/invidious/yt_backend/youtube_api.cr [hunk 2]
```diff
@@ -434,8 +425,27 @@ module YoutubeAPI
...
+    else
+      # Decompress the body ourselves, when using HTTP::Client given that
+      # auto-decompress is broken in the Crystal stdlib.
+      # Read more:
+      #  - https://github.com/iv-org/invidious/issues/2612
+      #  - https://github.com/crystal-lang/crystal/issues/11354
+      #
+      case response.headers["Content-Encoding"]?
+      when "gzip"
+        body = Compress::Gzip::Reader.new(response.body_io, sync_close: true)
```
Comment (syeopite) [inline src/invidious/yt_backend/youtube_api.cr:439]: Multiple encoding types can be specified in `Content-Encoding`. 

> ```
> // Multiple, in the order in which they were applied
> Content-Encoding: deflate, gzip
> ```

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding#syntax

Comment (SamantazFox) [reply src/invidious/yt_backend/youtube_api.cr:439]: The client may send multiple items, but the reply will only be in one format! The server won't return the content duplicated, once in gzip and another one in deflate.

**Comment by unixfox** [comment]
In testing on https://yewtu.be.
Waiting for the build: https://github.com/unixfox/invidious-custom/actions/runs/1473906192

File: src/invidious/yt_backend/youtube_api.cr (hunk 2 above)
Comment (syeopite) [reply src/invidious/yt_backend/youtube_api.cr:439]: No, what? Of course content isn't duplicated. The server can however return content that's encoded in two or more different formats, hence `Content-Encoding`. 

The above example is first compressed with `deflate` and then `gzip`. So in other to fetch the uncompressed content you'd need to decompress `gzip` and then decompress with `deflate`.

Comment (unixfox) [reply src/invidious/yt_backend/youtube_api.cr:439]: > The above example is first compressed with `deflate` and then `gzip`. So in other to fetch the uncompressed content you'd need to decompress `gzip` and then decompress with `deflate`.

Is the native HTTP client from Crystal really going to do that if it encounters a double compression?

Comment (syeopite) [reply src/invidious/yt_backend/youtube_api.cr:439]: Mozilla specifies that explicitly on their page for the `Content-Encoding` header, so I'd assume that is on the spec. Which means, stdlib HTTP::Client should most likely support that

**Comment by SamantazFox** [comment]
Ooops XD

**Comment by unixfox** [comment]
> Ooops XD

Not sure what happened, tho. Might be great to add to the CI a curl test to a video page.

File: src/invidious/yt_backend/youtube_api.cr (hunk 2 above)
Comment (SamantazFox) [reply src/invidious/yt_backend/youtube_api.cr:439]: > The above example is first compressed with `deflate` and then `gzip`. So in other to fetch the uncompressed content you'd need to decompress `gzip` and then decompress with `deflate`.

Oooooh! I didn't know that it was possible! I'll fix that, then.

> Is the native HTTP client from Crystal really going to do that if it encounters a double compression?

Well, no, the Crystal HTTP::Client doesn't support that. I copied the code above directly from the Crystal stdlib!

**Comment by SamantazFox** [comment]
in testing on test.invidious.io

**Comment by unixfox** [comment]
New bugs probably related to this PR:
- https://github.com/iv-org/invidious/issues/2664
- https://github.com/iv-org/invidious/issues/2663

**Comment by SamantazFox** [comment]
Merging, as the deployment on yewtu.be doesn't seem to cause issues anymore:
https://github.com/unixfox/invidious-custom/commit/b8daba935f69f41e97cfa2cdf82422cf70f4bb52

**Comment by LennyPenny** [comment]
now that this is merged I'm getting #2634 on my instance

**Comment by unixfox** [comment]
Now that I think of it, I only tested with quic enabled and not with HTTP1.

**Pull request merged**